src/index.js

```diff
@@ -35,169 +119,204 @@ class QueuedTask extends HasModels {
+   *   request and expected to return a promise that resolves to a task
+   *   state object; if omitted, the task is created with
+   *   `createTask()`, using the task name, and initializing `taskData`
+   *   to the contents of the request body
    */
-  async _requestHandler(route, initiate, req, res) {
+  async taskRequestRoute(route, taskName, initiate) {
+    if (!initiate)
+      initiate = (req) => {
+        let params = {name: taskName, taskData: req.body || {}}
+        return this.createTask(params)
+      }
+    users.protectedRoute(route)
+    await router.route('post', route, this._requestHandler.bind(this, route, taskName, initiate))
+  }
+
+  async _requestHandler(route, taskName, initiate, req, res) {
     let taskId = req.body.id,
         promise
 
-    function isChanged(task, timestamp) { return (!task || task.completed || (task.timestamp > timestamp)) }
+    function isChanged(state, timestamp) { return (!state || state.completed || (state.timestamp > timestamp)) }
 
     if (!taskId) { // initial request
-      let task
+      let state
       try {
-        task = await initiate(req)
-        task.route = route
-        await task.save()
+        state = await initiate(req)
+        state = await this._updateTaskState(state.id, {route})
       }
       catch (e) {
         this.log.error('queued-task _requestHandler ', e)
       }
-      promise = Promise.resolve(this._packageTask(task))
+      promise = Promise.resolve(state)
     }
-    else {
-      let task = await this.getTask(taskId)
-      if (route !== task.route) {
+    else { // status poll
+      let state = await this.getTaskState(taskId)
+      if (route !== state.route) {
         this.log.error('queued-task _requestHandler route mismatch')
         promise = Promise.resolve()
         res.status(400)
       }
       else {
         let timestamp = req.body.timestamp || 0
-        task = this._packageTask(task)
-        if (isChanged(task, timestamp))
-          promise = Promise.resolve(task)
+        if (isChanged(state, timestamp))
+          promise = Promise.resolve(state)
         else {
           let timerId,
               progressListener,
               timer = new Promise((resolve, reject) => {
                 timerId = setTimeout(() => {
-                  resolve(task)
+                  resolve(state)
                   timerId = undefined
                 }, taskProgressLimit)
               }),
               checker = new Promise((resolve, reject) => {
-                progressListener = (task) => {
-                  if (task.id === taskId) {
-                    task = this._packageTask(task)
-                    if (isChanged(task, timestamp)) resolve(task)
-                  }
+                progressListener = (state) => {
+                  if (state.id === taskId)
+                    if (isChanged(state, timestamp)) resolve(state)
                 }
                 queuedTask.on('queued-task-progress', progressListener)
```

Comment:
This whole bit feels sketchy, registering listeners for every task and then only acting on matched ids - would a simplification be to emit id-based events? (`queued-task-progress-${id}`) is there another global listener for these we'd want to still emit the app-wide event for?

Reply:
I don't know what the right pattern is for events. In the code I have now, I have per-queue listeners (so they filter on task name instead of task id). The single global event seemed simple and adequate, absent a clear direction on something more complicated.

There's the related issue of worker tasks persisting across restarts, so you may be getting events for a task that you've lost track of due to a restart.

And I could really use a tweak to `workerQueue` to gain direct access to the task queues, so I don't have to scrape the queue information from the job entry.

What I'd like to do is defer this, put it into an issue, and see what people's thoughts are.

Reply:
Yeah, later refactor in light of a future fix for direct access seems fine.